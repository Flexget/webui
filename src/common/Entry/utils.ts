import { RawEntry, CardType, FieldNames, MappingType, GettersType } from './types';
import { Entry, DefaultEntry } from './fields';
import {
  RawMovieEntry,
  MovieEntry,
  movieFieldList,
  MovieFields,
  MovieFieldNames,
} from './fields/movies';
import {
  RawSeriesEntry,
  SeriesEntry,
  seriesFieldList,
  SeriesFields,
  SeriesFieldNames,
} from './fields/series';

import {
  RawEpisodeEntry,
  EpisodeEntry,
  EpisodeFields,
  episodesFieldList,
  EpisodeFieldNames,
} from './fields/episodes';

const isMovie = (entry: RawEntry): entry is RawMovieEntry => !!entry.movieName;
const isSeries = (entry: RawEntry): entry is RawSeriesEntry =>
  !!entry.seriesName && !!entry.seriesSeason;
const isEpisode = (entry: RawEntry): entry is RawEpisodeEntry =>
  !!(entry.seriesSeason && entry.seriesEpisode);

const getPropFn = <U>(mapping: MappingType<U>) => {
  const mapByField = mapping.reduce<Partial<Record<FieldNames<U>, string[]>>>(
    (res, item) =>
      Object.entries(item).reduce(
        (out, [k, v]) => ({
          ...out,
          [k]: [...(out[k] ?? []), v],
        }),
        res,
      ),
    {},
  );
  return (entry: U, field: FieldNames<U>) =>
    (mapByField[field] ?? ([] as string[])).reduce((res, val) => res ?? entry[val], undefined);
};

const makeGetters = <T>(names: Record<string, FieldNames<T>>, mapping: MappingType<T>) => {
  const getProp = getPropFn<T>(mapping);
  return (entry: T) =>
    Object.values(names).reduce(
      (obj, key: FieldNames<T>) => ({
        ...obj,
        get [key]() {
          return getProp(entry, key);
        },
      }),
      {} as GettersType<T>,
    );
};

const getMovieProps = makeGetters<MovieFields>(MovieFieldNames, movieFieldList);
const getSeriesProps = makeGetters<SeriesFields>(SeriesFieldNames, seriesFieldList);
const getEpisodeProps = makeGetters<EpisodeFields>(EpisodeFieldNames, episodesFieldList);

const toMovieEntry = (entry: RawMovieEntry): MovieEntry => {
  return {
    ...entry,
    type: CardType.Movie,
    ...getMovieProps(entry),
  };
};

const toSeriesEntry = (entry: RawSeriesEntry): SeriesEntry => ({
  ...entry,
  type: CardType.Series,
  ...getSeriesProps(entry),
});

const toEpisodeEntry = (entry: RawEpisodeEntry): EpisodeEntry => ({
  ...entry,
  type: CardType.Episode,
  ...getEpisodeProps(entry),
});

const toDefaultEntry = (entry: RawEntry): DefaultEntry => ({
  ...entry,
  type: CardType.Default,
});

export const toEntry = (entry: RawEntry): Entry => {
  if (isMovie(entry)) {
    return toMovieEntry(entry);
  }
  if (isSeries(entry)) {
    return toSeriesEntry(entry);
  }
  if (isEpisode(entry)) {
    return toEpisodeEntry(entry);
  }
  return toDefaultEntry(entry);
};
