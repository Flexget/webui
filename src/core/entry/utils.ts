import { RawEntry, FieldNames, MappingType, GettersType, BaseEntry } from './types';
import { MovieEntry, movieFieldList, RawMovieEntry, MovieFieldNames } from './fields/movies';
import { SeriesEntry, seriesFieldList, SeriesFieldNames, RawSeriesEntry } from './fields/series';
import {
  EpisodeEntry,
  RawEpisodeEntry,
  episodesFieldList,
  EpisodeFieldNames,
} from './fields/episodes';

export const isMovie = (entry: RawEntry): entry is RawMovieEntry => !!entry.movieName;
export const isSeries = (entry: RawEntry): entry is RawSeriesEntry =>
  !!entry.seriesName && !entry.seriesSeason;
export const isEpisode = (entry: RawEntry): entry is RawEpisodeEntry & SeriesEntry =>
  !!(entry.seriesSeason && entry.seriesEpisode && entry.seriesName);

const getPropFn = <U>(mapping: MappingType<U>) => {
  const mapByField = mapping.reduce<Partial<Record<FieldNames<U>, string[]>>>(
    (res, item) =>
      Object.entries(item).reduce(
        (out, [k, v]) => ({
          ...out,
          [k]: [...(out[k] ?? []), v],
        }),
        res,
      ),
    {},
  );
  return (entry: U, field: FieldNames<U>) =>
    (mapByField[field] ?? ([] as string[])).reduce((res, val) => res ?? entry[val], undefined);
};

const makeGetters = <T>(names: Record<string, FieldNames<T>>, mapping: MappingType<T>) => {
  const getProp = getPropFn<T>(mapping);
  return (entry: T) =>
    Object.values(names).reduce(
      (obj, key: FieldNames<T>) => ({
        ...obj,
        get [key]() {
          return getProp(entry, key);
        },
      }),
      {} as GettersType<T>,
    );
};

const getMovieProps = makeGetters<RawMovieEntry>(MovieFieldNames, movieFieldList);
const getSeriesProps = makeGetters<RawSeriesEntry>(SeriesFieldNames, seriesFieldList);
const getEpisodeProps = makeGetters<RawEpisodeEntry>(EpisodeFieldNames, episodesFieldList);

export const toMovieEntry = (entry: RawMovieEntry): MovieEntry => {
  return {
    ...entry,
    ...getMovieProps(entry),
  };
};

export const toSeriesEntry = (entry: RawSeriesEntry): SeriesEntry => ({
  ...entry,
  ...getSeriesProps(entry),
});

export const toEpisodeEntry = (entry: RawEpisodeEntry): EpisodeEntry => ({
  ...entry,
  ...getEpisodeProps(entry),
});

export const toDefaultEntry = (entry: BaseEntry): BaseEntry => ({
  ...entry,
});
