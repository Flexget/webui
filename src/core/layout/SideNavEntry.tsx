import React, { ComponentType, FC } from 'react';
import { css } from '@emotion/core';
import { Link } from 'react-router-dom';
import { ListItem, ListItemText, ListItemIcon } from '@material-ui/core';
import theme from 'theme';

interface Props {
  onClick?: () => void;
  path?: string;
  Icon: ComponentType;
  className?: string;
  name: string;
}

export const colorClass = css`
  color: ${theme.palette.secondary[200]};
`;

export const navItem = css`
  ${colorClass};
  border-left: 3px solid transparent;
  cursor: pointer;

  &:hover {
    border-left: 3px solid ${theme.palette.primary[500]};
  }
`;

const SideNavEntry: FC<Props> = ({ onClick, path, Icon, name, className }) => {
  const item = (
    <ListItem css={[navItem, className]} onClick={onClick}>
      <ListItemIcon css={colorClass}>
        <Icon />
      </ListItemIcon>
      <ListItemText css={colorClass} disableTypography primary={name} />
    </ListItem>
  );

  if (path) {
    return <Link to={path}>{item}</Link>;
  }

  return item;
};

export default SideNavEntry;
