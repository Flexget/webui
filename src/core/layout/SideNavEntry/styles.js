import styled from '@emotion/styled';
import { css } from '@emotion/core';
import theme from 'theme';
import ListItem from '@material-ui/core/ListItem';

export const colorClass = css`
  color: ${theme.palette.secondary[200]};
`;

export const NavItem = styled(ListItem)`
  border-left: 3px solid transparent;
  cursor: pointer;

  &:hover {
    border-left: 3px solid ${theme.palette.primary[500]};
  }
`;
