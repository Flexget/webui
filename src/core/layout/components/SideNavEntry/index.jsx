import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { ListItemIcon } from 'material-ui/List';
import {
  SideNavIcon,
  SideNavText,
  NavItem,
} from './styles';

const SideNavEntry = ({ onClick, path, icon, name, className }) => {
  const item = (
    <NavItem className={className} onClick={onClick}>
      <ListItemIcon><SideNavIcon icon={icon} fixedWidth /></ListItemIcon>
      <SideNavText disableTypography inset primary={name} />
    </NavItem>
  );

  if (path) {
    return <Link to={path}>{item}</Link>;
  }

  return item;
};

SideNavEntry.propTypes = {
  onClick: PropTypes.func.isRequired,
  path: PropTypes.string,
  icon: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  className: PropTypes.string,
};

SideNavEntry.defaultProps = {
  className: '',
  path: '',
};

export default SideNavEntry;
