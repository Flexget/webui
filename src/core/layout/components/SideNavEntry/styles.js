import styled, { css } from 'react-emotion';
import theme from 'theme';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';

const colorClass = css`color: ${theme.palette.secondary[200]};`;

export const SideNavIcon = styled(FontAwesomeIcon)`
  ${colorClass};
`;

export const SideNavText = styled(ListItemText)`
  ${colorClass};
`;

export const NavItem = styled(ListItem)`
  border-left: 3px solid transparent;
  cursor: pointer;

  &:hover {
    border-left: 3px solid ${theme.palette.primary[500]};
  }
`;
