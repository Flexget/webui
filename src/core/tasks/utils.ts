import { RawEntry } from 'core/entry/types';

export const toExecuteRequest = (entry?: RawEntry) => {
  if (entry) {
    const { id, title, originalUrl, ...fields } = entry;
    return { title, url: originalUrl, fields };
  }
  return { title: '', url: '', fields: {} };
};
