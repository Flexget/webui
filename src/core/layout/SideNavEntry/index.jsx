import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import { NavItem, colorClass } from './styles';

const SideNavEntry = ({ onClick, path, Icon, name, className }) => {
  const item = (
    <NavItem css={[colorClass, className]} onClick={onClick}>
      <ListItemIcon css={colorClass}>
        <Icon />
      </ListItemIcon>
      <ListItemText css={colorClass} disableTypography primary={name} />
    </NavItem>
  );

  if (path) {
    return <Link to={path}>{item}</Link>;
  }

  return item;
};

SideNavEntry.propTypes = {
  onClick: PropTypes.func.isRequired,
  path: PropTypes.string,
  Icon: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  className: PropTypes.string,
};

SideNavEntry.defaultProps = {
  className: '',
  path: '',
};

export default SideNavEntry;
