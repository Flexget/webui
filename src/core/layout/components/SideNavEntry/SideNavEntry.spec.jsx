import React from 'react';
import renderer from 'react-test-renderer';
import SideNavEntry from 'core/layout/components/SideNavEntry';
import { themed, router } from 'utils/tests';

describe('core/layout/components/SideNavEntry', () => {
  it('renders correctly with link', () => {
    const tree = renderer.create(router(themed(<SideNavEntry icon="icon" label="Test" link="/test" />))).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly without link', () => {
    const tree = renderer.create(themed(<SideNavEntry icon="icon" label="Test" />)).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
