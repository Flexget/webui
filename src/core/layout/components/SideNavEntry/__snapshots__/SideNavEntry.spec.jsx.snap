// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/layout/components/SideNavEntry renders correctly with link 1`] = `
.glamor-1 {
  border-left: 3px solid transparent;
  cursor: pointer;
}

.glamor-1:hover {
  border-left: 3px solid #ff9800;
}

.glamor-0 {
  color: #b0bec5;
}

<li
  className="MuiListItem-root-1 MuiListItem-default-4 MuiListItem-gutters-8 glamor-1 e1vhtnjr2"
  disabled={false}
>
  <div
    className="MuiListItemText-root-12 MuiListItemText-inset-13 glamor-0 e1vhtnjr1"
  />
</li>
`;

exports[`core/layout/components/SideNavEntry renders correctly without link 1`] = `
.glamor-1 {
  border-left: 3px solid transparent;
  cursor: pointer;
}

.glamor-1:hover {
  border-left: 3px solid #ff9800;
}

.glamor-0 {
  color: #b0bec5;
}

<li
  className="MuiListItem-root-1 MuiListItem-default-4 MuiListItem-gutters-8 glamor-1 e1vhtnjr2"
  disabled={false}
>
  <div
    className="MuiListItemText-root-12 MuiListItemText-inset-13 glamor-0 e1vhtnjr1"
  />
</li>
`;
